File: packages/sdk-py/dna/adapters/search/sqlite_vec.py
"""SqliteVecRecordSearchProvider — the embeddable default RecordSearchProvider.

rsh-memory-similarity-evolution → rec-embeddable-provider. The FIRST real
implementation of the ``RecordSearchProvider`` port (the kernel shipped only
the port + an honest lexical fallback). Hybrid search entirely inside one
SQLite file per scope, offline, no server:

  * dense plane   — sqlite-vec ``vec0`` KNN over ``kernel.embed()`` vectors
                    (the deterministic ``FakeEmbeddingProvider`` floor by
                    default; any registered ``EmbeddingPort`` transparently).
  * lexical plane — FTS5 BM25 over the same text.
  * fusion        — Reciprocal Rank Fusion (``rrf.reciprocal_rank_fusion``, a
                    pure function shared with the TS twin and any future
                    pgvector provider).

The store schema is OWNED by the shared migration contract (``migrations.py`` +
``adapters/_migrations.run_migrations``) — closing f-embeddings-ddl-debt.

Overlay/tenant-aware: one ``.db`` per scope, ``tenant`` a column ('' = base).
A tenant search reads base ∪ overlay and the overlay row shadows the base row
for the same ``(kind, name)`` — mirroring the source reader's overlay merge.

Parity: the TS twin is ``src/adapters/search/sqlite-vec.ts`` (same tables, same
RRF, same overlay-shadow); both are exercised by the shared
``record_search_conformance_suite`` with the fake embedder, offline.
"""
from __future__ import annotations

import hashlib
import logging
import os
import sqlite3
import struct
from pathlib import Path
from typing import Any, Iterable

from dna.adapters.search.rrf import DEFAULT_RRF_K, reciprocal_rank_fusion

logger = logging.getLogger(__name__)

#: How many candidates to pull from each plane before fusing + filtering. The
#: over-fetch absorbs attrition from the kind/tenant/overlay filter (which runs
#: AFTER ranking) so the final top-k is stable. Scales with the requested k.
_OVERFETCH = 4
_MIN_CANDIDATES = 40


def _load_sqlite_vec(conn: sqlite3.Connection) -> None:
    """Load the sqlite-vec loadable extension into ``conn``. The ``sqlite-vec``
    pip package bundles the platform extension + a ``load`` helper."""
    try:
        import sqlite_vec
    except ImportError as exc:  # pragma: no cover — exercised via the extra
        raise ImportError(
            "SqliteVecRecordSearchProvider needs the 'search-sqlite' extra: "
            "pip install 'dna-sdk[search-sqlite]'"
        ) from exc
    conn.enable_load_extension(True)
    sqlite_vec.load(conn)
    conn.enable_load_extension(False)


def _serialize_f32(vector: Iterable[float]) -> bytes:
    """Pack a float vector into sqlite-vec's compact float32 blob layout."""
    vec = list(vector)
    return struct.pack(f"{len(vec)}f", *vec)


def document_text(raw: dict[str, Any]) -> str:
    """Derive the searchable text blob for a raw document.

    Walks the doc's ``spec`` (and top-level ``metadata.name``) collecting every
    STRING value, in document order — the same "string values only" discipline
    the kernel's lexical fallback uses, so dense/lexical see the same corpus.
    Callers may override by passing an explicit ``text`` to ``index``.
    """
    parts: list[str] = []
    meta = raw.get("metadata") or {}
    name = meta.get("name") or raw.get("name")
    if isinstance(name, str):
        parts.append(name)

    def _walk(node: Any) -> None:
        if isinstance(node, str):
            parts.append(node)
        elif isinstance(node, dict):
            for value in node.values():
                _walk(value)
        elif isinstance(node, (list, tuple)):
            for value in node:
                _walk(value)

    _walk(raw.get("spec") or {})
    return "\n".join(p for p in parts if p)


class SqliteVecRecordSearchProvider:
    """``RecordSearchProvider`` backed by sqlite-vec + FTS5 + RRF.

    Construct with the kernel (for ``embed``) and either a ``db_dir`` (one
    ``<scope>.db`` per scope, the production shape) or an explicit ``db_path``
    (a single file — handy for tests). Register via
    ``kernel.record_search_provider(provider)``.
    """

    def __init__(
        self,
        kernel: Any,
        *,
        db_dir: str | os.PathLike[str] | None = None,
        db_path: str | os.PathLike[str] | None = None,
        rrf_k: int = DEFAULT_RRF_K,
    ) -> None:
        if db_dir is None and db_path is None:
            db_dir = os.getenv("DNA_SEARCH_DIR") or ".dna-search"
        self._kernel = kernel
        self._db_dir = Path(db_dir) if db_dir is not None else None
        self._single_path = Path(db_path) if db_path is not None else None
        self._rrf_k = rrf_k
        self._conns: dict[str, sqlite3.Connection] = {}

    # ------------------------------------------------------------------
    # store / schema (migration-owned)
    # ------------------------------------------------------------------

    def _path_for(self, scope: str) -> Path:
        if self._single_path is not None:
            return self._single_path
        assert self._db_dir is not None
        self._db_dir.mkdir(parents=True, exist_ok=True)
        safe = scope.replace(os.sep, "_").replace("/", "_")
        return self._db_dir / f"{safe}.db"

    async def _conn_for(self, scope: str) -> sqlite3.Connection:
        """Open (once per store path) a connection with sqlite-vec loaded and
        the schema migrated + identity-pinned. Cached for the provider's life.

        Async because the shared migration runner is async-shaped; the sqlite3
        work inside the callbacks is synchronous and stays on this thread/loop
        (no cross-thread connection use)."""
        path = str(self._path_for(scope))
        conn = self._conns.get(path)
        if conn is not None:
            return conn
        conn = sqlite3.connect(path)
        conn.row_factory = sqlite3.Row
        _load_sqlite_vec(conn)
        await self._migrate(conn)
        self._pin_identity(conn)
        self._conns[path] = conn
        return conn

    async def _migrate(self, conn: sqlite3.Connection) -> list[int]:
        """Apply pending migrations through the shared forward-only runner —
        the search store's schema has a real owner (closes f-embeddings-ddl-debt)."""
        from dna.adapters._migrations import run_migrations
        from dna.adapters.search.migrations import build_migrations

        dims = int(self._kernel.embedding_dims)
        migrations = build_migrations(dims)

        async def ensure_control_table() -> None:
            conn.execute(
                "CREATE TABLE IF NOT EXISTS schema_migrations "
                "(version INTEGER PRIMARY KEY, applied_at TEXT NOT NULL)"
            )
            conn.commit()

        async def fetch_applied() -> list[int]:
            return [r["version"] for r in conn.execute(
                "SELECT version FROM schema_migrations"
            ).fetchall()]

        async def apply_version(version: int, script: str) -> None:
            conn.executescript(script)
            conn.execute(
                "INSERT INTO schema_migrations (version, applied_at) VALUES (?, ?)",
                (version, _now()),
            )
            conn.commit()

        return await run_migrations(
            migrations,
            ensure_control_table=ensure_control_table,
            fetch_applied=fetch_applied,
            apply_version=apply_version,
            dialect="SQLite(search)",
        )

    def _pin_identity(self, conn: sqlite3.Connection) -> None:
        """Refuse to reuse a store built for a different embedding space —
        mixing vectors from different (dims, model_id) is silently wrong."""
        dims = str(int(self._kernel.embedding_dims))
        model_id = str(self._kernel.embedding_model_id)
        rows = {r["key"]: r["value"] for r in conn.execute(
            "SELECT key, value FROM search_meta"
        ).fetchall()}
        if not rows:
            conn.executemany(
                "INSERT INTO search_meta (key, value) VALUES (?, ?)",
                [("embedding_dims", dims), ("embedding_model_id", model_id)],
            )
            conn.commit()
            return
        if rows.get("embedding_dims") != dims or rows.get("embedding_model_id") != model_id:
            raise ValueError(
                "search store was built for embedding space "
                f"({rows.get('embedding_model_id')}, dims={rows.get('embedding_dims')}) "
                f"but the active provider is ({model_id}, dims={dims}) — "
                "the vectors are incomparable. Use a fresh store dir or re-index."
            )

    # ------------------------------------------------------------------
    # index / delete
    # ------------------------------------------------------------------

    async def index(self, records: list[dict[str, Any]]) -> int:
        """Index (upsert) records into their scope's store.

        Each record: ``{scope, kind, name, tenant?, text?, raw?, title?,
        snippet?}``. ``text`` is used verbatim if present; otherwise it is
        derived from ``raw`` via :func:`document_text`. Idempotent by text
        hash — re-indexing unchanged text is skipped (no re-embed). Returns
        the number of records actually (re)embedded.
        """
        if not records:
            return 0
        # Group by scope so each store is touched once; embed in one batch.
        pending: list[tuple[dict[str, Any], str, str]] = []  # (rec, text, hash)
        for rec in records:
            text = rec.get("text")
            if text is None:
                text = document_text(rec.get("raw") or {})
            text = text or ""
            h = hashlib.sha256(text.encode("utf-8")).hexdigest()
            pending.append((rec, text, h))

        # Decide which need (re)embedding — skip rows whose hash is unchanged.
        to_embed: list[tuple[dict[str, Any], str, str]] = []
        for rec, text, h in pending:
            conn = await self._conn_for(rec["scope"])
            existing = conn.execute(
                "SELECT text_hash FROM search_docs "
                "WHERE scope=? AND kind=? AND name=? AND tenant=?",
                (rec["scope"], rec["kind"], rec["name"], rec.get("tenant") or ""),
            ).fetchone()
            if existing is not None and existing["text_hash"] == h:
                continue
            to_embed.append((rec, text, h))

        if not to_embed:
            return 0

        vectors = await self._kernel.embed([t for _, t, _ in to_embed])
        for (rec, text, h), vector in zip(to_embed, vectors):
            self._upsert(rec, text, h, vector)
        for scope in {rec["scope"] for rec, _, _ in to_embed}:
            self._cached_conn(scope).commit()
        return len(to_embed)

    def _cached_conn(self, scope: str) -> sqlite3.Connection:
        """The already-open connection for ``scope`` (opened by ``_conn_for``
        earlier in the same async call). Sync helpers use this so they never
        need to await."""
        return self._conns[str(self._path_for(scope))]

    def _upsert(
        self, rec: dict[str, Any], text: str, text_hash: str, vector: list[float],
    ) -> None:
        conn = self._cached_conn(rec["scope"])
        scope, kind, name = rec["scope"], rec["kind"], rec["name"]
        tenant = rec.get("tenant") or ""
        title = rec.get("title")
        snippet = rec.get("snippet") or _snippet(text)
        # Upsert metadata; capture the stable rowid for the virtual tables.
        row = conn.execute(
            "SELECT rowid FROM search_docs "
            "WHERE scope=? AND kind=? AND name=? AND tenant=?",
            (scope, kind, name, tenant),
        ).fetchone()
        if row is None:
            cur = conn.execute(
                "INSERT INTO search_docs "
                "(scope, kind, name, tenant, text_hash, title, snippet, text) "
                "VALUES (?,?,?,?,?,?,?,?)",
                (scope, kind, name, tenant, text_hash, title, snippet, text),
            )
            rowid = cur.lastrowid
        else:
            rowid = row["rowid"]
            conn.execute(
                "UPDATE search_docs SET text_hash=?, title=?, snippet=?, text=? "
                "WHERE rowid=?",
                (text_hash, title, snippet, text, rowid),
            )
            conn.execute("DELETE FROM search_vec WHERE doc_rowid=?", (rowid,))
            conn.execute("DELETE FROM search_fts WHERE rowid=?", (rowid,))
        conn.execute(
            "INSERT INTO search_vec (doc_rowid, embedding) VALUES (?, ?)",
            (rowid, _serialize_f32(vector)),
        )
        conn.execute(
            "INSERT INTO search_fts (rowid, text) VALUES (?, ?)", (rowid, text),
        )

    async def delete(self, ids: list[dict[str, Any] | tuple]) -> int:
        """Delete indexed records. Each id is a dict ``{scope, kind, name,
        tenant?}`` or a ``(scope, kind, name[, tenant])`` tuple. Returns the
        number of rows removed."""
        removed = 0
        touched: set[str] = set()
        for ident in ids:
            if isinstance(ident, dict):
                scope, kind, name = ident["scope"], ident["kind"], ident["name"]
                tenant = ident.get("tenant") or ""
            else:
                scope, kind, name = ident[0], ident[1], ident[2]
                tenant = ident[3] if len(ident) > 3 and ident[3] else ""
            conn = await self._conn_for(scope)
            row = conn.execute(
                "SELECT rowid FROM search_docs "
                "WHERE scope=? AND kind=? AND name=? AND tenant=?",
                (scope, kind, name, tenant),
            ).fetchone()
            if row is None:
                continue
            rowid = row["rowid"]
            conn.execute("DELETE FROM search_vec WHERE doc_rowid=?", (rowid,))
            conn.execute("DELETE FROM search_fts WHERE rowid=?", (rowid,))
            conn.execute("DELETE FROM search_docs WHERE rowid=?", (rowid,))
            removed += 1
            touched.add(scope)
        for scope in touched:
            self._cached_conn(scope).commit()
        return removed

    # ------------------------------------------------------------------
    # search (RecordSearchProvider)
    # ------------------------------------------------------------------

    async def search(
        self, *, scope: str, query_text: str, kind: str | None = None,
        k: int = 10, tenant: str = "",
    ) -> list[dict[str, Any]]:
        """Hybrid dense+lexical search fused with RRF. Returns hits shaped
        ``{scope, kind, name, score, title?, snippet?, rank_dense?,
        rank_lexical?}`` (the port's guaranteed keys plus optional extras),
        ordered best-first. Overlay-aware: base ∪ overlay, overlay shadows base."""
        if not query_text.strip() or k <= 0:
            return []
        conn = await self._conn_for(scope)
        overfetch = max(_MIN_CANDIDATES, k * _OVERFETCH)

        # Dense plane — vec0 KNN over the query embedding.
        query_vec = (await self._kernel.embed([query_text]))[0]
        dense_ranked: list[int] = []
        if any(query_vec):  # all-zero query (no tokens) → skip dense plane
            dense_ranked = [r["doc_rowid"] for r in conn.execute(
                "SELECT doc_rowid FROM search_vec "
                "WHERE embedding MATCH ? ORDER BY distance LIMIT ?",
                (_serialize_f32(query_vec), overfetch),
            ).fetchall()]

        # Lexical plane — FTS5 BM25. An unparseable MATCH query is not an error;
        # it just contributes no lexical ranks.
        lexical_ranked: list[int] = []
        fts_query = _fts_query(query_text)
        if fts_query:
            try:
                lexical_ranked = [r["rowid"] for r in conn.execute(
                    "SELECT rowid FROM search_fts "
                    "WHERE search_fts MATCH ? ORDER BY bm25(search_fts) LIMIT ?",
                    (fts_query, overfetch),
                ).fetchall()]
            except sqlite3.OperationalError:
                lexical_ranked = []

        if not dense_ranked and not lexical_ranked:
            return []

        # Fuse ranks (RRF is pure + string-keyed → stringify rowids).
        fused = reciprocal_rank_fusion(
            [[str(r) for r in dense_ranked], [str(r) for r in lexical_ranked]],
            k=self._rrf_k,
        )
        dense_pos = {r: i + 1 for i, r in enumerate(dense_ranked)}
        lexical_pos = {r: i + 1 for i, r in enumerate(lexical_ranked)}

        # Resolve metadata + apply kind/tenant filter with overlay shadowing.
        rowids = [int(rid) for rid, _ in fused]
        meta = self._resolve_meta(conn, rowids)
        best: dict[tuple[str, str], dict[str, Any]] = {}
        for rid, score in fused:
            rowid = int(rid)
            m = meta.get(rowid)
            if m is None:
                continue
            if kind is not None and m["kind"] != kind:
                continue
            row_tenant = m["tenant"] or ""
            if row_tenant not in ("", tenant or ""):
                continue  # a different tenant's overlay — never leaks
            key = (m["kind"], m["name"])
            prev = best.get(key)
            # Overlay (matching tenant) shadows base (''); otherwise higher
            # fused score wins.
            if prev is None:
                best[key] = _hit(scope, m, score, dense_pos.get(rowid), lexical_pos.get(rowid))
            else:
                prev_is_base = prev["_tenant"] == ""
                this_is_overlay = row_tenant != "" and row_tenant == (tenant or "")
                if this_is_overlay and prev_is_base:
                    best[key] = _hit(scope, m, score, dense_pos.get(rowid), lexical_pos.get(rowid))

        hits = sorted(best.values(), key=lambda h: -h["score"])
        for h in hits:
            h.pop("_tenant", None)
        return hits[:k]

    def _resolve_meta(
        self, conn: sqlite3.Connection, rowids: list[int],
    ) -> dict[int, sqlite3.Row]:
        if not rowids:
            return {}
        placeholders = ",".join("?" * len(rowids))
        rows = conn.execute(
            f"SELECT rowid, scope, kind, name, tenant, title, snippet "
            f"FROM search_docs WHERE rowid IN ({placeholders})",
            rowids,
        ).fetchall()
        return {r["rowid"]: r for r in rows}

    # ------------------------------------------------------------------
    # lifecycle
    # ------------------------------------------------------------------

    def close(self) -> None:
        for conn in self._conns.values():
            try:
                conn.close()
            except Exception:  # noqa: BLE001
                pass
        self._conns.clear()


# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------

def _now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def _snippet(text: str, max_len: int = 200) -> str:
    flat = " ".join(text.split())
    return flat[:max_len] + ("…" if len(flat) > max_len else "")


def _hit(
    scope: str, m: sqlite3.Row, score: float,
    rank_dense: int | None, rank_lexical: int | None,
) -> dict[str, Any]:
    hit: dict[str, Any] = {
        "scope": scope, "kind": m["kind"], "name": m["name"], "score": score,
        "_tenant": m["tenant"] or "",
    }
    if m["title"]:
        hit["title"] = m["title"]
    if m["snippet"]:
        hit["snippet"] = m["snippet"]
    if rank_dense is not None:
        hit["rank_dense"] = rank_dense
    if rank_lexical is not None:
        hit["rank_lexical"] = rank_lexical
    return hit


def _fts_query(query_text: str) -> str:
    """Turn free text into a safe FTS5 MATCH query: OR of the alphanumeric
    tokens, each quoted so FTS5 never interprets a token as an operator."""
    import re
    tokens = re.findall(r"[a-z0-9]+", query_text.lower())
    return " OR ".join(f'"{t}"' for t in tokens)

File: packages/sdk-ts/src/adapters/search/sqlite-vec.ts
/**
 * SqliteVecRecordSearchProvider — the embeddable default RecordSearchProvider
 * (TS twin of `dna/adapters/search/sqlite_vec.py`).
 *
 * The FIRST real implementation of the `RecordSearchProvider` port on the TS
 * side. Hybrid search inside one SQLite file per scope, offline, no server:
 *
 *   - dense   — sqlite-vec `vec0` KNN over `kernel.embed()` vectors (the
 *               deterministic `FakeEmbeddingProvider` floor by default).
 *   - lexical — FTS5 BM25 over the same text.
 *   - fusion  — Reciprocal Rank Fusion (`reciprocalRankFusion`, the pure
 *               function shared with the Python twin).
 *
 * The store schema is OWNED by a numbered migration (`migrations.ts`), closing
 * f-embeddings-ddl-debt. Overlay/tenant-aware: `tenant` is a column ('' = base);
 * a tenant search reads base ∪ overlay and the overlay row shadows the base row
 * for the same `(kind, name)`.
 *
 * SQLite access goes through {@link openSearchDb} (bun:sqlite or node:sqlite);
 * sqlite-vec is an OPTIONAL peer dep (`sqlite-vec`), never in the default graph.
 * Both are exercised by the shared record-search conformance kit with the fake
 * embedder, offline.
 */
import { sha256 } from "js-sha256";

import { DEFAULT_RRF_K, reciprocalRankFusion } from "./rrf.js";
import { buildMigrations, runSearchMigrations } from "./migrations.js";
import { openSearchDb, serializeF32, type SearchDb } from "./driver.js";

/** The slice of the kernel this provider needs (embedding surface). */
export interface EmbeddingKernel {
  embed(texts: string[]): Promise<number[][]>;
  readonly embeddingDims: number;
  readonly embeddingModelId: string;
}

/** A record to index. `text` is used verbatim if present; otherwise derived
 *  from `raw` via {@link documentText}. */
export interface SearchRecord {
  scope: string;
  kind: string;
  name: string;
  tenant?: string;
  text?: string;
  raw?: Record<string, unknown>;
  title?: string;
  snippet?: string;
}

export interface SearchHit extends Record<string, unknown> {
  scope: string;
  kind: string;
  name: string;
  score: number;
}

/** An id to delete: `{ scope, kind, name, tenant? }`. */
export interface RecordId {
  scope: string;
  kind: string;
  name: string;
  tenant?: string;
}

const OVERFETCH = 4;
const MIN_CANDIDATES = 40;

/** Derive the searchable text blob for a raw document — every string value in
 *  `spec` plus `metadata.name`, in document order (parity with the Python
 *  `document_text`). */
export function documentText(raw: Record<string, unknown>): string {
  const parts: string[] = [];
  const meta = (raw.metadata as Record<string, unknown> | undefined) ?? {};
  const name = (meta.name as string | undefined) ?? (raw.name as string | undefined);
  if (typeof name === "string") parts.push(name);
  const walk = (node: unknown): void => {
    if (typeof node === "string") parts.push(node);
    else if (Array.isArray(node)) node.forEach(walk);
    else if (node && typeof node === "object") Object.values(node).forEach(walk);
  };
  walk(raw.spec ?? {});
  return parts.filter((p) => p).join("\n");
}

export class SqliteVecRecordSearchProvider {
  private readonly kernel: EmbeddingKernel;
  private readonly dbDir: string | null;
  private readonly singlePath: string | null;
  private readonly rrfK: number;
  private readonly conns = new Map<string, SearchDb>();

  constructor(
    kernel: EmbeddingKernel,
    opts: { dbDir?: string; dbPath?: string; rrfK?: number } = {},
  ) {
    this.kernel = kernel;
    let dbDir = opts.dbDir ?? null;
    if (dbDir === null && !opts.dbPath) {
      const env = (globalThis as { process?: { env?: Record<string, string> } })
        .process?.env;
      dbDir = env?.DNA_SEARCH_DIR ?? ".dna-search";
    }
    this.dbDir = opts.dbPath ? null : dbDir;
    this.singlePath = opts.dbPath ?? null;
    this.rrfK = opts.rrfK ?? DEFAULT_RRF_K;
  }

  // ------------------------------------------------------------------
  // store / schema (migration-owned)
  // ------------------------------------------------------------------

  private pathFor(scope: string): string {
    if (this.singlePath) return this.singlePath;
    const safe = scope.replace(/[/\\]/g, "_");
    return `${this.dbDir}/${safe}.db`;
  }

  private async connFor(scope: string): Promise<SearchDb> {
    const path = this.pathFor(scope);
    const existing = this.conns.get(path);
    if (existing) return existing;
    if (this.dbDir && !this.singlePath) await ensureDir(this.dbDir);
    const db = await openSearchDb(path);
    runSearchMigrations(db, buildMigrations(this.kernel.embeddingDims));
    this.pinIdentity(db);
    this.conns.set(path, db);
    return db;
  }

  private pinIdentity(db: SearchDb): void {
    const dims = String(this.kernel.embeddingDims);
    const modelId = this.kernel.embeddingModelId;
    const rows = db.all<{ key: string; value: string }>(
      "SELECT key, value FROM search_meta",
    );
    if (rows.length === 0) {
      db.run("INSERT INTO search_meta (key, value) VALUES (?, ?)", ["embedding_dims", dims]);
      db.run("INSERT INTO search_meta (key, value) VALUES (?, ?)", ["embedding_model_id", modelId]);
      return;
    }
    const meta = Object.fromEntries(rows.map((r) => [r.key, r.value]));
    if (meta.embedding_dims !== dims || meta.embedding_model_id !== modelId) {
      throw new Error(
        `search store was built for embedding space (${meta.embedding_model_id}, `
          + `dims=${meta.embedding_dims}) but the active provider is (${modelId}, `
          + `dims=${dims}) — the vectors are incomparable. Use a fresh store dir `
          + "or re-index.",
      );
    }
  }

  // ------------------------------------------------------------------
  // index / delete
  // ------------------------------------------------------------------

  /** Index (upsert) records into their scope's store. Idempotent by text hash
   *  (re-indexing unchanged text is skipped). Returns the count actually
   *  (re)embedded. */
  async index(records: SearchRecord[]): Promise<number> {
    if (records.length === 0) return 0;
    const pending = records.map((rec) => {
      const text = rec.text ?? documentText(rec.raw ?? {});
      return { rec, text, hash: sha256(text) };
    });

    const toEmbed: typeof pending = [];
    for (const item of pending) {
      const db = await this.connFor(item.rec.scope);
      const existing = db.get<{ text_hash: string }>(
        "SELECT text_hash FROM search_docs WHERE scope=? AND kind=? AND name=? AND tenant=?",
        [item.rec.scope, item.rec.kind, item.rec.name, item.rec.tenant ?? ""],
      );
      if (existing && existing.text_hash === item.hash) continue;
      toEmbed.push(item);
    }
    if (toEmbed.length === 0) return 0;

    const vectors = await this.kernel.embed(toEmbed.map((i) => i.text));
    toEmbed.forEach((item, i) => this.upsert(item.rec, item.text, item.hash, vectors[i]!));
    return toEmbed.length;
  }

  private upsert(
    rec: SearchRecord, text: string, textHash: string, vector: number[],
  ): void {
    const db = this.conns.get(this.pathFor(rec.scope))!;
    const tenant = rec.tenant ?? "";
    const title = rec.title ?? null;
    const snippet = rec.snippet ?? snippetOf(text);
    const row = db.get<{ rowid: number }>(
      "SELECT rowid FROM search_docs WHERE scope=? AND kind=? AND name=? AND tenant=?",
      [rec.scope, rec.kind, rec.name, tenant],
    );
    let rowid: number;
    if (!row) {
      const info = db.run(
        "INSERT INTO search_docs (scope, kind, name, tenant, text_hash, title, snippet, text) "
          + "VALUES (?,?,?,?,?,?,?,?)",
        [rec.scope, rec.kind, rec.name, tenant, textHash, title, snippet, text],
      );
      rowid = info.lastInsertRowid;
    } else {
      rowid = row.rowid;
      db.run("UPDATE search_docs SET text_hash=?, title=?, snippet=?, text=? WHERE rowid=?", [
        textHash, title, snippet, text, rowid,
      ]);
      db.run("DELETE FROM search_vec WHERE doc_rowid=?", [BigInt(rowid)]);
      db.run("DELETE FROM search_fts WHERE rowid=?", [BigInt(rowid)]);
    }
    // vec0 (and fts5) require an INTEGER rowid binding — node:sqlite binds a JS
    // number as a double and vec0 rejects it, so pass BigInt (both drivers OK).
    db.run("INSERT INTO search_vec (doc_rowid, embedding) VALUES (?, ?)", [
      BigInt(rowid), serializeF32(vector),
    ]);
    db.run("INSERT INTO search_fts (rowid, text) VALUES (?, ?)", [BigInt(rowid), text]);
  }

  /** Delete indexed records. Returns the number of rows removed. */
  async delete(ids: RecordId[]): Promise<number> {
    let removed = 0;
    for (const id of ids) {
      const db = await this.connFor(id.scope);
      const row = db.get<{ rowid: number }>(
        "SELECT rowid FROM search_docs WHERE scope=? AND kind=? AND name=? AND tenant=?",
        [id.scope, id.kind, id.name, id.tenant ?? ""],
      );
      if (!row) continue;
      const rowid = BigInt(row.rowid);
      db.run("DELETE FROM search_vec WHERE doc_rowid=?", [rowid]);
      db.run("DELETE FROM search_fts WHERE rowid=?", [rowid]);
      db.run("DELETE FROM search_docs WHERE rowid=?", [rowid]);
      removed += 1;
    }
    return removed;
  }

  // ------------------------------------------------------------------
  // search (RecordSearchProvider)
  // ------------------------------------------------------------------

  async search(opts: {
    scope: string;
    queryText: string;
    kind?: string | null;
    k?: number;
    tenant?: string;
  }): Promise<SearchHit[]> {
    const { scope, queryText } = opts;
    const kind = opts.kind ?? null;
    const k = opts.k ?? 10;
    const tenant = opts.tenant ?? "";
    if (!queryText.trim() || k <= 0) return [];
    const db = await this.connFor(scope);
    const overfetch = Math.max(MIN_CANDIDATES, k * OVERFETCH);

    const queryVec = (await this.kernel.embed([queryText]))[0]!;
    let denseRanked: number[] = [];
    if (queryVec.some((x) => x !== 0)) {
      denseRanked = db
        .all<{ doc_rowid: number }>(
          "SELECT doc_rowid FROM search_vec WHERE embedding MATCH ? ORDER BY distance LIMIT ?",
          [serializeF32(queryVec), overfetch],
        )
        .map((r) => r.doc_rowid);
    }

    let lexicalRanked: number[] = [];
    const ftsQuery = ftsQueryOf(queryText);
    if (ftsQuery) {
      try {
        lexicalRanked = db
          .all<{ rowid: number }>(
            "SELECT rowid FROM search_fts WHERE search_fts MATCH ? ORDER BY bm25(search_fts) LIMIT ?",
            [ftsQuery, overfetch],
          )
          .map((r) => r.rowid);
      } catch {
        lexicalRanked = [];
      }
    }
    if (denseRanked.length === 0 && lexicalRanked.length === 0) return [];

    const fused = reciprocalRankFusion(
      [denseRanked.map(String), lexicalRanked.map(String)],
      this.rrfK,
    );
    const densePos = new Map(denseRanked.map((r, i) => [r, i + 1]));
    const lexicalPos = new Map(lexicalRanked.map((r, i) => [r, i + 1]));

    const rowids = fused.map(([rid]) => Number(rid));
    const meta = this.resolveMeta(db, rowids);
    const best = new Map<string, SearchHit & { _tenant: string }>();
    for (const [rid, score] of fused) {
      const rowid = Number(rid);
      const m = meta.get(rowid);
      if (!m) continue;
      if (kind !== null && m.kind !== kind) continue;
      const rowTenant = m.tenant ?? "";
      if (rowTenant !== "" && rowTenant !== (tenant || "")) continue; // other tenant — never leaks
      const key = `${m.kind} ${m.name}`;
      const prev = best.get(key);
      const hit = makeHit(scope, m, score, densePos.get(rowid), lexicalPos.get(rowid));
      if (!prev) {
        best.set(key, hit);
      } else {
        const thisIsOverlay = rowTenant !== "" && rowTenant === (tenant || "");
        if (thisIsOverlay && prev._tenant === "") best.set(key, hit);
      }
    }
    const hits = [...best.values()].sort((a, b) => b.score - a.score);
    for (const h of hits) delete (h as { _tenant?: string })._tenant;
    return hits.slice(0, k);
  }

  private resolveMeta(db: SearchDb, rowids: number[]): Map<number, MetaRow> {
    if (rowids.length === 0) return new Map();
    const placeholders = rowids.map(() => "?").join(",");
    const rows = db.all<MetaRow>(
      `SELECT rowid, scope, kind, name, tenant, title, snippet FROM search_docs `
        + `WHERE rowid IN (${placeholders})`,
      rowids,
    );
    return new Map(rows.map((r) => [r.rowid, r]));
  }

  close(): void {
    for (const db of this.conns.values()) {
      try {
        db.close();
      } catch {
        /* ignore */
      }
    }
    this.conns.clear();
  }
}

interface MetaRow {
  rowid: number;
  scope: string;
  kind: string;
  name: string;
  tenant: string | null;
  title: string | null;
  snippet: string | null;
}

function makeHit(
  scope: string, m: MetaRow, score: number,
  rankDense?: number, rankLexical?: number,
): SearchHit & { _tenant: string } {
  const hit: SearchHit & { _tenant: string } = {
    scope, kind: m.kind, name: m.name, score, _tenant: m.tenant ?? "",
  };
  if (m.title) hit.title = m.title;
  if (m.snippet) hit.snippet = m.snippet;
  if (rankDense !== undefined) hit.rank_dense = rankDense;
  if (rankLexical !== undefined) hit.rank_lexical = rankLexical;
  return hit;
}

function snippetOf(text: string, maxLen = 200): string {
  const flat = text.split(/\s+/).filter(Boolean).join(" ");
  return flat.length > maxLen ? flat.slice(0, maxLen) + "…" : flat;
}

/** OR of quoted alphanumeric tokens — parity with the Python `_fts_query`. */
function ftsQueryOf(queryText: string): string {
  const tokens = queryText.toLowerCase().match(/[a-z0-9]+/g) ?? [];
  return tokens.map((t) => `"${t}"`).join(" OR ");
}

async function ensureDir(dir: string): Promise<void> {
  try {
    const fs = (await import("node:fs")) as {
      mkdirSync(p: string, o: { recursive: boolean }): void;
    };
    fs.mkdirSync(dir, { recursive: true });
  } catch {
    /* best effort — open will surface a real failure */
  }
}
